tests/test_visualizer.py

```diff
@@ -19,19 +19,21 @@ def test_visualizer_run_end(input_data, target_data, perturbation, tmp_path):
     target_list = [target_data]
 
     # simulate an addition perturbation
-    def perturb(input):
+    def perturb(input, target):
         result = [sample + perturbation for sample in input]
-        return result
+        return result, target
 
+    adversary = Mock(spec=Adversary, side_effect=perturb)
     trainer = Mock()
-    model = Mock(composer=Mock(return_value=perturb(input_list)))
     outputs = Mock()
-    batch = {"input": input_list, "target": target_list}
-    adversary = Mock(spec=Adversary, side_effect=perturb)
+    target_model = Mock()
+
+    # Canonical batch in Adversary.
+    batch = (input_list, target_list, target_model)
 
     visualizer = PerturbedImageVisualizer(folder)
-    visualizer.on_train_batch_end(trainer, model, outputs, batch, 0)
-    visualizer.on_train_end(trainer, model)
+    visualizer.on_train_batch_end(trainer, adversary, outputs, batch, 0)
+    visualizer.on_train_end(trainer, adversary)
```

Comment:
Does this mean this no longer works with universal perturbations?

Reply:
Actually, this writes to disk (which is a bad idea). I wish this wrote to tensorboard. I swear there was a PR that did that...

Reply:
OK. I will revisit the visualizer later when I have a chance.